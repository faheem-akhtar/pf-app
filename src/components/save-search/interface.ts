import { SaveSearchFrequencyEnum } from 'enums/save-search/frequency.enum';

import { SaveSearchFiltersInterface } from './filters.interface';

export interface SaveSearchInterface {
  /**
   * ID
   */
  id: string;

  /**
   * Name of the saved search
   */
  name: string;

  /**
   * Frequency to which email will be sent: Weekly | daily | off
   */
  frequency: SaveSearchFrequencyEnum;

  /**
   * Filters in use for the current saved search
   */
  filters: SaveSearchFiltersInterface;

  /**
   * Formatted name of the search according to selected filters
   */
  formattedFilters: string;
}
