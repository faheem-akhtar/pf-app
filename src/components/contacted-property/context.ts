import { createContext } from 'react';

import { ContactedPropertyContextInterface } from './context.interface';

export const ContactedPropertyContext = createContext<ContactedPropertyContextInterface>({
  data: [],
  add: () => null,
});
