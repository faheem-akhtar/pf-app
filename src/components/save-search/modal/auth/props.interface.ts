export interface SaveSearchModalAuthPropsInterface {
  /**
   * Call on success login or register
   */
  onSuccess: () => void;

  /**
   * Call if operation is cancelled
   */
  onCancel: () => void;
}
