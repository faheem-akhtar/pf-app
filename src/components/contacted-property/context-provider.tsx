import { FunctionComponent } from 'react';

import { apiContactedPropertiesCreateFetcher } from 'api/contacted-properties/create/fetcher';
import { useApiContactedProperties } from 'api/contacted-properties/hook';
import { useUserPropertyState } from 'components/user-property/state.hook';
import { dateToIso } from 'helpers/date/to-iso';

import { ContactedPropertyContext } from './context';
import { ContactedPropertyContextInterface } from './context.interface';
import { ContactedPropertyInterface } from './interface';
import { CONTACTED_PROPERTY_STORAGE_KEY } from './storage-key.constant';

export const ContactedPropertyContextProvider: FunctionComponent = ({ children }) => {
  const contactedPropertiesResponse = useApiContactedProperties();
  const [contactedProperties, setContactedProperty] = useUserPropertyState<ContactedPropertyInterface>(
    CONTACTED_PROPERTY_STORAGE_KEY,
    contactedPropertiesResponse,
    ({ propertyId, contactType, contactDate }) =>
      ({ propertyId, contactType, contactDate } as ContactedPropertyInterface),
    apiContactedPropertiesCreateFetcher
  );

  const value: ContactedPropertyContextInterface = {
    data: contactedProperties,
    add: (propertyId, contactType) => {
      const property = {
        propertyId,
        contactType,
        contactDate: dateToIso(new Date()),
      } as ContactedPropertyInterface;

      setContactedProperty(property);
    },
  };

  return <ContactedPropertyContext.Provider value={value}>{children}</ContactedPropertyContext.Provider>;
};
