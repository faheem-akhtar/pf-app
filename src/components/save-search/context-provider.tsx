import { FunctionComponent, useContext, useEffect, useState } from 'react';

import { apiSaveSearchCreateFetcher } from 'api/save-search/create/fetcher';
import { useApiSaveSearch } from 'api/save-search/hook';
import { FiltersContext } from 'components/filters/context';
import { UserContext } from 'components/user/context';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';

import { SaveSearchContext } from './context';
import { SaveSearchContextInterface } from './context.interface';
import { saveSearchFilterEquality } from './filter-equality';
import { saveSearchFiltersMapper } from './filters-mapper';
import { SaveSearchInterface } from './interface';
import { saveSearchTracker } from './tracker';

export const SaveSearchContextProvider: FunctionComponent = ({ children }) => {
  const user = useContext(UserContext);
  const saveSearchResponse = useApiSaveSearch();
  const saveSearchResponseData = saveSearchResponse.ok ? saveSearchResponse.data : [];
  const [searches, setSearches] = useState<SaveSearchInterface[]>([]);
  const filtersCtx = useContext(FiltersContext);
  const filters = saveSearchFiltersMapper(filtersCtx?.value || {});

  useEffect(() => {
    if (user && saveSearchResponse.ok) {
      setSearches(saveSearchResponseData);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user, saveSearchResponse.ok]);

  const value: SaveSearchContextInterface = {
    ok: saveSearchResponse.ok,
    data: searches,
    filtered: searches.filter((item) => saveSearchFilterEquality(item.filters, filters)),
    create: ({ name, frequency }) =>
      apiSaveSearchCreateFetcher({ name, frequency, filters: filtersCtx.value }).then((response) => {
        if (response.ok) {
          setSearches([...searches, response.data]);
          saveSearchTracker.onCreateSuccess(
            filtersCtx.value[FiltersParametersEnum.locationsIds].map((location) => location.id)
          );
        }
        return response;
      }),
  };

  return <SaveSearchContext.Provider value={value}>{children}</SaveSearchContext.Provider>;
};
