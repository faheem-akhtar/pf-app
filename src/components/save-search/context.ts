import { createContext } from 'react';
import { SaveSearchContextInterface } from './context.interface';
import { SaveSearchLoadResultInterface } from './load-result-interface';

export const SaveSearchContext = createContext<SaveSearchContextInterface>({
  data: [],
  filtered: [],
  create: () => Promise.resolve({ ok: true, data: {} as SaveSearchLoadResultInterface, headers: {} as Headers }),
});
