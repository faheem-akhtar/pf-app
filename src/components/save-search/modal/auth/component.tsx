import React, { FunctionComponent, useContext, useEffect, useRef } from 'react';

import { AuthModalComponent } from 'components/auth/modal/component';
import { ModalComponent } from 'components/modal/component';
import { SaveSearchContext } from 'components/save-search/context';
import { saveSearchTracker } from 'components/save-search/tracker';
import { UserContext } from 'components/user/context';
import { functionNoop } from 'helpers/function/noop';
import { AuthService } from 'services/auth/service';
import { AuthSubscribeEventTypeEnum } from 'services/auth/subscribe-event-type.enum';
import { AuthSubscriberType } from 'services/auth/subscriber.type';

import { SaveSearchModalAuthPropsInterface } from './props.interface';

export const SaveSearchModalAuthComponent: FunctionComponent<SaveSearchModalAuthPropsInterface> = ({
  onCancel,
  onSuccess,
}) => {
  const user = useContext(UserContext);
  const saveSearch = useContext(SaveSearchContext);

  const openLoginRef = useRef(functionNoop);
  const closeLoginRef = useRef(functionNoop);

  useEffect(() => {
    const onUpdate: AuthSubscriberType = (newUser, meta) => {
      if (newUser && meta?.eventType === AuthSubscribeEventTypeEnum.login) {
        saveSearchTracker.onSignInSuccess();
      }
      if (newUser && meta?.eventType === AuthSubscribeEventTypeEnum.register) {
        saveSearchTracker.onSignUpSuccess();
      }
    };
    const unsubscribeAuth = AuthService.subscribe(onUpdate);

    openLoginRef.current();

    return (): void => unsubscribeAuth();
  });

  useEffect(() => {
    if (user && saveSearch.ok) {
      onSuccess();
    }
  }, [user, saveSearch.ok, onSuccess]);

  return (
    <ModalComponent openRef={openLoginRef} closeRef={closeLoginRef} overlay>
      <AuthModalComponent cancel={onCancel} />
    </ModalComponent>
  );
};
