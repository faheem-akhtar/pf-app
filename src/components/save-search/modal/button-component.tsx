import { useContext, useEffect, useRef, useState } from 'react';

import { AuthModalComponent } from 'components/auth/modal/component';
import { IconThinStarTemplate } from 'components/icon/thin/star-template';
import { ModalComponent } from 'components/modal/component';
import { SnackbarContext } from 'components/snackbar/context';
import { UserContext } from 'context/user/context';
import { useTranslation } from 'helpers/translation/hook';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { OnBoardingComponent } from 'library/on-boarding/component';
import { OnBoardingTypeEnum } from 'library/on-boarding/type.enum';
import { AuthService } from 'services/auth/service';
import { AuthSubscribeEventTypeEnum } from 'services/auth/subscribe-event-type.enum';
import { AuthSubscriberType } from 'services/auth/subscriber.type';

import { SaveSearchContext } from '../context';
import { saveSearchTracker } from '../tracker';
import { SaveSearchModalContentComponent } from './content-component';
import styles from './save-search-modal-component.module.scss';

export const SaveSearchModalButtonComponent = ({ visibleTooltip }: { visibleTooltip: boolean }): JSX.Element => {
  const { t } = useTranslation();
  const user = useContext(UserContext);
  const snackbar = useContext(SnackbarContext);
  const saveSearch = useContext(SaveSearchContext);
  const [autoOpenDialog, setAutoOpenDialog] = useState(false);
  const openFiltersRef = useRef<() => void>(() => null);
  const closeFiltersRef = useRef<() => void>(() => null);
  const openLoginRef = useRef<() => void>(() => null);
  const closeLoginRef = useRef<() => void>(() => null);

  const openDialog = (): void => {
    if (user) {
      openFiltersRef.current();
      saveSearchTracker.onOpenCreateDialog();
    } else {
      openLoginRef.current();
      setAutoOpenDialog(true);
    }
  };

  useEffect(() => {
    if (autoOpenDialog && user) {
      closeLoginRef.current();
      if (saveSearch.filtered.length === 0) {
        openDialog();
      } else {
        snackbar.alert({ message: t('save_search/exist-notification') });
      }
      setAutoOpenDialog(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [autoOpenDialog, user, saveSearch.filtered]);

  useEffect(() => {
    if (autoOpenDialog) {
      const onUpdate: AuthSubscriberType = (user, meta) => {
        if (user && meta?.eventType === AuthSubscribeEventTypeEnum.login) {
          saveSearchTracker.onSignInSuccess();
        }
        if (user && meta?.eventType === AuthSubscribeEventTypeEnum.register) {
          saveSearchTracker.onSignUpSuccess();
        }
      };
      const unsubscribeAuth = AuthService.subscribe(onUpdate);

      return (): void => unsubscribeAuth();
    }
  }, [autoOpenDialog]);

  return (
    <>
      <div className={styles.tooltipWrapper}>
        <ButtonTemplate
          type='button'
          componentType={ButtonComponentTypeEnum.tertiary}
          size={ButtonSizeEnum.small}
          onClick={(): void => {
            if (saveSearch.filtered.length === 0) {
              openDialog();
            }
            saveSearchTracker.onClickCta();
          }}
          icon={{
            component: IconThinStarTemplate,
            position: ButtonIconPositionEnum.left,
            className: saveSearch.filtered.length ? styles.icon__active : styles.icon__inactive,
          }}
        >
          {t('save-search/cta-label')}
        </ButtonTemplate>
        {visibleTooltip && (
          <OnBoardingComponent
            name={OnBoardingTypeEnum.saveSearchTooltip}
            className={styles.tooltip}
            arrowClassName={styles.tooltipArrow}
            onClose={(): void => {
              saveSearchTracker.onTooltipOpen();
              saveSearchTracker.onTooltipClose();
            }}
          >
            <>
              <p className={styles.tooltipTitle}>{t('on-boarding/save-search-title')}</p>
              {t('on-boarding/save-search-description')}
            </>
          </OnBoardingComponent>
        )}
      </div>
      <ModalComponent openRef={openFiltersRef} closeRef={closeFiltersRef}>
        <SaveSearchModalContentComponent close={(): void => closeFiltersRef.current()} />
      </ModalComponent>
      <ModalComponent openRef={openLoginRef} closeRef={closeLoginRef} overlay>
        <AuthModalComponent
          cancel={(): void => {
            closeLoginRef.current();
            setAutoOpenDialog(false);
          }}
        />
      </ModalComponent>
    </>
  );
};
