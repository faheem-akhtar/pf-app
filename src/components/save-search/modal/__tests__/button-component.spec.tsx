/**
 * @jest-environment jsdom
 */

import { act, render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { mockModalEnv } from 'mocks/modal-env/mock';
import { mockReactUseSwr } from 'mocks/react/use-swr.mock';
import { userModelStub } from 'stubs/user/model.stub';

import { UserContext } from 'context/user/context';
import { UserContextProvider } from 'context/user/context-provider';
import { AuthService } from 'services/auth/service';
import { AuthSubscribeEventTypeEnum } from 'services/auth/subscribe-event-type.enum';
import { UserModelInterface } from 'services/user/model.interface';
import { WindowService } from 'services/window/service';
import { AnalyticsGaEventType } from 'types/analytics/ga/event.type';

import { SaveSearchModalButtonComponent } from '../button-component';

jest.mock('services/window/service');

jest.mock('../content-component', () => ({
  // eslint-disable-next-line react/display-name
  SaveSearchModalContentComponent: ({ close }: { close: () => void }): JSX.Element => (
    <button data-testid='save-search-modal-content-component' onClick={close} />
  ),
}));

jest.mock('components/auth/modal/component', () => ({
  // eslint-disable-next-line react/display-name
  AuthModalComponent: ({ cancel }: { cancel: () => void }): JSX.Element => (
    <button data-testid='auth-modal-component' onClick={cancel} />
  ),
}));

describe('SaveSearchModalButtonComponent', () => {
  beforeEach(() => {
    (global as unknown as { dataLayer: AnalyticsGaEventType[] }).dataLayer = [];
    document.documentElement.scroll = jest.fn();

    mockReactUseSwr('api_user', {
      ok: true,
      data: {},
    });

    mockModalEnv();
  });

  it('should render the component', () => {
    expect(render(<SaveSearchModalButtonComponent visibleTooltip />).container).toMatchSnapshot();
  });

  describe('customer', () => {
    let renderResult: RenderResult;

    beforeEach(() => {
      renderResult = render(
        <UserContext.Provider value={userModelStub()}>
          <SaveSearchModalButtonComponent visibleTooltip />
        </UserContext.Provider>
      );
    });

    it('should open and close the save search modal when clicked on the cta', () => {
      act(() => {
        userEvent.click(renderResult.getByText('save-search/cta-label'));
      });

      expect(renderResult.queryByTestId('save-search-modal-content-component')).toBeInTheDocument();

      act(() => {
        userEvent.click(renderResult.getByTestId('save-search-modal-content-component'));
      });

      expect(renderResult.queryByTestId('save-search-modal-content-component')).not.toBeInTheDocument();
    });
  });

  describe('guest', () => {
    let renderResult: RenderResult;

    beforeEach(() => {
      (WindowService.localStorage.getItem as jest.Mock).mockReturnValue(null);

      renderResult = render(
        <UserContextProvider>
          <SaveSearchModalButtonComponent visibleTooltip />
        </UserContextProvider>
      );
    });

    it('should open and close the login dialog', () => {
      act(() => {
        userEvent.click(renderResult.getByText('save-search/cta-label'));
      });

      expect(renderResult.queryByTestId('auth-modal-component')).toBeInTheDocument();

      act(() => {
        userEvent.click(renderResult.getByTestId('auth-modal-component'));
      });

      expect(renderResult.queryByTestId('auth-modal-component')).not.toBeInTheDocument();
    });

    it('should open the save search modal when user got logged in', () => {
      act(() => {
        userEvent.click(renderResult.getByText('save-search/cta-label'));
      });

      expect(renderResult.queryByTestId('auth-modal-component')).toBeInTheDocument();

      expect((global as unknown as { dataLayer: AnalyticsGaEventType[] }).dataLayer).toEqual(
        expect.arrayContaining([
          {
            event: 'customEvent',
            eventAction: 'click',
            eventCategory: 'Saved Search',
          },
        ])
      );

      act(() => {
        AuthService.onAuthResolved(
          {
            user: { userId: '2' } as UserModelInterface,
            meta: { token: '', refresh_token: '' },
          },
          AuthSubscribeEventTypeEnum.login
        );
      });

      expect((global as unknown as { dataLayer: AnalyticsGaEventType[] }).dataLayer).toEqual(
        expect.arrayContaining([
          {
            event: 'customEvent',
            eventAction: 'signIn',
            eventCategory: 'Saved Search',
          },
        ])
      );
    });

    it('should trigger signUp event when user register while using the save search cta', () => {
      act(() => {
        userEvent.click(renderResult.getByText('save-search/cta-label'));
      });

      expect((global as unknown as { dataLayer: AnalyticsGaEventType[] }).dataLayer).toEqual(
        expect.not.arrayContaining([
          {
            event: 'customEvent',
            eventAction: 'signUp',
            eventCategory: 'Saved Search',
          },
        ])
      );

      act(() => {
        AuthService.onAuthResolved(
          {
            user: { userId: '3' } as UserModelInterface,
            meta: { token: '', refresh_token: '' },
          },
          AuthSubscribeEventTypeEnum.register
        );
      });

      expect((global as unknown as { dataLayer: AnalyticsGaEventType[] }).dataLayer).toEqual(
        expect.arrayContaining([
          {
            event: 'customEvent',
            eventAction: 'signUp',
            eventCategory: 'Saved Search',
          },
        ])
      );
    });
  });
});
