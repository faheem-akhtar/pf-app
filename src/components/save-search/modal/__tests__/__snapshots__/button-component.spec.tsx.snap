// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveSearchModalButtonComponent should render the component 1`] = `
<div>
  <div
    class="tooltipWrapper"
  >
    <button
      class="button tertiary small"
    >
      <svg
        class="icon icon--left icon__inactive"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M11.5516 3.03827C11.735 2.66665 12.265 2.66665 12.4484 3.03827L14.5518 7.30025C14.7703 7.74296 15.1926 8.04982 15.6812 8.12081L20.3846 8.80425C20.7947 8.86384 20.9584 9.36783 20.6617 9.65709L17.2583 12.9746C16.9047 13.3192 16.7434 13.8157 16.8269 14.3023L17.6303 18.9866C17.7004 19.3951 17.2717 19.7066 16.9048 19.5137L12.698 17.3021C12.261 17.0723 11.739 17.0723 11.302 17.3021L7.09516 19.5137C6.72835 19.7066 6.29963 19.3951 6.36969 18.9866L7.17312 14.3023C7.25658 13.8157 7.09525 13.3192 6.74173 12.9746L3.33833 9.6571C3.04158 9.36783 3.20533 8.86384 3.61544 8.80425L8.31881 8.12081C8.80738 8.04982 9.22973 7.74296 9.44822 7.30025L11.5516 3.03827Z"
          stroke="currentColor"
        />
      </svg>
      save-search/cta-label
    </button>
    <div
      class="container tooltip container--bottom container--visible"
    >
      <div
        class="content"
      >
        <div
          class="arrow tooltipArrow"
        />
        <div
          class="inner"
          role="tooltip"
        >
          <span>
            <p
              class="tooltipTitle"
            >
              on-boarding/save-search-title
            </p>
            on-boarding/save-search-description
          </span>
        </div>
        <div
          class="action"
        >
          <svg
            class="icon"
            viewBox="3 3 10 10"
          >
            <path
              clip-rule="evenodd"
              d="M4.34408 3.17475C4.01717 2.92057 3.54448 2.94367 3.24408 3.24408C2.91864 3.56951 2.91864 4.09715 3.24408 4.42259L6.82149 8L3.24408 11.5774C2.91864 11.9028 2.91864 12.4305 3.24408 12.7559C3.54448 13.0563 4.01717 13.0794 4.34408 12.8252L4.42259 12.7559L8 9.17851L11.5774 12.7559L11.6559 12.8252C11.9828 13.0794 12.4555 13.0563 12.7559 12.7559C13.0814 12.4305 13.0814 11.9028 12.7559 11.5774L9.17851 8L12.7559 4.42259C13.0814 4.09715 13.0814 3.56951 12.7559 3.24408C12.4555 2.94367 11.9828 2.92057 11.6559 3.17475L11.5774 3.24408L8 6.82149L4.42259 3.24408L4.34408 3.17475Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
